File: PixelogicKit/Sources/PixelogicKit/Scoring.swift
// Pure scoring model (port of src/engine/scoring.ts).
//
// Per-puzzle score (0–100):  round(100 × min(1, par/bestTime)) − assist penalties.
// Pixelogic Score (0–1600):  difficulty- and badge-weighted average of best
//                            per-puzzle scores across the whole library.

import Foundation

extension Difficulty {
    /// How much a tier contributes to the overall rating. Harder ⇒ worth far more.
    public var weight: Double {
        switch self {
        case .easy: 1
        case .medium: 2
        case .hard: 4
        case .expert: 7
        case .max: 12
        }
    }

    /// Target ("par") seconds-per-cell by tier.
    var secondsPerCell: Double {
        switch self {
        case .easy: 1.0
        case .medium: 1.5
        case .hard: 2.2
        case .expert: 3.2
        case .max: 4.5
        }
    }

    /// How many "Check square" reveals a tier allows per attempt (nil = unlimited).
    public var checkBudget: Int? {
        switch self {
        case .easy, .medium: nil
        case .hard: 3
        case .expert: 2
        case .max: 1
        }
    }
}

/// Par time for a puzzle, in whole seconds.
public func parSeconds(_ difficulty: Difficulty, area: Int) -> Int {
    Int((Double(area) * difficulty.secondsPerCell).rounded())
}

/// Assist usage during a single solve attempt.
public struct AssistTally: Codable, Sendable, Equatable {
    public var checkSquare: Int = 0
    public var checkLine: Int = 0
    public var checkBoard: Int = 0
    public var hint: Int = 0
    /// Fill-out or Watch-solve was used — the attempt scores 0.
    public var voided: Bool = false

    public init() {}

    public static let penaltyCheckSquare = 5
    public static let penaltyCheckLine = 15
    public static let penaltyCheckBoard = 40
    public static let penaltyHint = 20

    public var penaltyTotal: Int {
        checkSquare * Self.penaltyCheckSquare
            + checkLine * Self.penaltyCheckLine
            + checkBoard * Self.penaltyCheckBoard
            + hint * Self.penaltyHint
    }
}

/// Per-puzzle score in [0,100]. Hitting par (or faster) gives full speed credit.
public func puzzleScore(
    difficulty: Difficulty, area: Int, bestTimeMs: Int, assists: AssistTally
) -> Int {
    if assists.voided || bestTimeMs <= 0 { return 0 }
    let par = Double(parSeconds(difficulty, area: area))
    let speed = min(1.0, par / (Double(bestTimeMs) / 1000.0))
    let raw = (100.0 * speed).rounded() - Double(assists.penaltyTotal)
    return max(0, min(100, Int(raw)))
}

public struct PuzzleMeta: Sendable {
    public let id: String
    public let difficulty: Difficulty
    /// Badge weight multiplier (<1 for easier-badge puzzles, >1 for harder ones).
    public let weightMult: Double

    public init(id: String, difficulty: Difficulty, weightMult: Double = 1) {
        self.id = id
        self.difficulty = difficulty
        self.weightMult = weightMult
    }
}

/// Overall Pixelogic Score in [0,1600]. Unsolved puzzles count as 0; badge
/// multipliers shift each puzzle's share (a perfect run is 1600 regardless).
public func pixelogicScore(bestScores: [String: Int], library: [PuzzleMeta]) -> Int {
    var earned = 0.0
    var possible = 0.0
    for p in library {
        let w = p.difficulty.weight * p.weightMult
        possible += w
        earned += w * (Double(bestScores[p.id] ?? 0) / 100.0)
    }
    return possible > 0 ? Int((1600.0 * earned / possible).rounded()) : 0
}

/// A flavour title for a Pixelogic Score, shown beneath the laurel.
public func scoreTitle(_ score: Int) -> String {
    switch score {
    case 1500...: "Grandmaster"
    case 1350...: "Master"
    case 1100...: "Expert"
    case 850...: "Sharp"
    case 550...: "Solver"
    case 250...: "Apprentice"
    default: "Novice"
    }
}

File: src/engine/scoring.ts
// Pure scoring model (DOM-free, unit-tested).
//
// Per-puzzle score (0–100):  round(100 × min(1, par/bestTime)) − assist penalties.
// Clueweave Score (0–1600):  difficulty-weighted average of best per-puzzle scores
//                            across the whole library (a 1600 = perfect everywhere).

import type { Difficulty } from "./types";

/** How much a tier contributes to the overall rating. Harder ⇒ worth far more. */
export const DIFFICULTY_WEIGHT: Record<Difficulty, number> = {
  easy: 1,
  medium: 2,
  hard: 4,
  expert: 7,
  max: 12,
};

/** Target ("par") seconds-per-cell by tier — the pace a solid solver should hit. */
const SECONDS_PER_CELL: Record<Difficulty, number> = {
  easy: 1.0,
  medium: 1.5,
  hard: 2.2,
  expert: 3.2,
  max: 4.5,
};

/** Par time for a puzzle, in whole seconds. */
export function parSeconds(difficulty: Difficulty, area: number): number {
  return Math.round(area * SECONDS_PER_CELL[difficulty]);
}

/** Assist usage during a single solve attempt. */
export interface AssistTally {
  checkSquare: number;
  checkLine: number;
  checkBoard: number;
  hint: number;
  /** Fill-out or Watch-solve was used — the attempt scores 0. */
  voided: boolean;
}

export const PENALTY = { checkSquare: 5, checkLine: 15, checkBoard: 40, hint: 20 } as const;

export function emptyTally(): AssistTally {
  return { checkSquare: 0, checkLine: 0, checkBoard: 0, hint: 0, voided: false };
}

export function penaltyTotal(a: AssistTally): number {
  return (
    a.checkSquare * PENALTY.checkSquare +
    a.checkLine * PENALTY.checkLine +
    a.checkBoard * PENALTY.checkBoard +
    a.hint * PENALTY.hint
  );
}

/** Per-puzzle score in [0,100]. Hitting par (or faster) gives full speed credit. */
export function puzzleScore(opts: {
  difficulty: Difficulty;
  area: number;
  bestTimeMs: number;
  assists: AssistTally;
}): number {
  if (opts.assists.voided || opts.bestTimeMs <= 0) return 0;
  const par = parSeconds(opts.difficulty, opts.area);
  const speed = Math.min(1, par / (opts.bestTimeMs / 1000));
  return Math.max(0, Math.min(100, Math.round(100 * speed - penaltyTotal(opts.assists))));
}

export interface PuzzleMeta {
  id: string;
  difficulty: Difficulty;
  /** Badge weight multiplier (<1 for easier-badge puzzles, >1 for harder ones). */
  weightMult?: number;
}

/**
 * Overall Clueweave Score in [0,1600]: a difficulty-weighted fraction of the
 * total possible score across `library`. Unsolved puzzles count as 0. Badge
 * multipliers shift how much each puzzle is worth relative to the others —
 * the perfect total stays 1600 because both sides scale together.
 */
export function clueweaveScore(bestScores: Record<string, number>, library: PuzzleMeta[]): number {
  let earned = 0;
  let possible = 0;
  for (const p of library) {
    const w = DIFFICULTY_WEIGHT[p.difficulty] * (p.weightMult ?? 1);
    possible += w;
    earned += w * ((bestScores[p.id] ?? 0) / 100);
  }
  return possible > 0 ? Math.round((1600 * earned) / possible) : 0;
}

/** How many "Check square" reveals a tier allows per attempt (∞ on easy/medium). */
export function checkBudget(difficulty: Difficulty): number {
  switch (difficulty) {
    case "hard":
      return 3;
    case "expert":
      return 2;
    case "max":
      return 1;
    default:
      return Infinity;
  }
}

/** A flavour title for a Clueweave Score, shown beneath the laurel. */
export function scoreTitle(score: number): string {
  if (score >= 1500) return "Grandmaster";
  if (score >= 1350) return "Master";
  if (score >= 1100) return "Expert";
  if (score >= 850) return "Sharp";
  if (score >= 550) return "Solver";
  if (score >= 250) return "Apprentice";
  return "Novice";
}
